types/src/metered_channel.rs

```diff
@@ -0,0 +1,295 @@
+use std::task::{Context, Poll};
+use tokio::sync::mpsc::{
+    self,
+    error::{SendError, TryRecvError, TrySendError},
+    OwnedPermit, Permit,
+};
+
+/// An [`mpsc::Sender`](tokio::sync::mpsc::Sender) with an [`IntGauge`]
+/// counting the number of currently queued items.
+#[derive(Debug)]
+pub struct Sender<T> {
+    inner: mpsc::Sender<T>,
+    gauge: IntGauge,
+}
+
+impl<T> Clone for Sender<T> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner.clone(),
+            gauge: self.gauge.clone(),
+        }
+    }
+}
+
+/// An [`mpsc::Receiver`](tokio::sync::mpsc::Receiver) with an [`IntGauge`]
+/// counting the number of currently queued items.
+#[derive(Debug)]
+pub struct Receiver<T> {
+    inner: mpsc::Receiver<T>,
+    gauge: IntGauge,
+}
+
+impl<T> Receiver<T> {
+    /// Receives the next value for this receiver.
+    /// Decrements the gauge in case of a successful `recv`.
+    pub async fn recv(&mut self) -> Option<T> {
+        self.inner.recv().inspect(|_| self.gauge.dec()).await
+    }
+
+    /// Attempts to receive the next value for this receiver.
+    /// Decrements the gauge in case of a successful `try_recv`.
+    pub async fn try_recv(&mut self) -> Result<Option<T>, TryRecvError> {
+        self.inner
+            .recv()
+            .map(|val| {
+                self.gauge.dec();
+                Ok(val)
+            })
+            .await
+    }
+
+    // TODO: facade [`blocking_recv`](tokio::mpsc::Receiver::blocking_recv) under the tokio feature flag "sync"
+
+    /// Closes the receiving half of a channel without dropping it.
+    pub fn close(&mut self) {
+        self.inner.close()
+    }
+
+    /// Polls to receive the next message on this channel.
+    ///  Decrements the gauge in case of a successful `poll_recv`.
+    pub fn poll_recv(&mut self, cx: &mut Context<'_>) -> Poll<Option<T>> {
+        match self.inner.poll_recv(cx) {
+            Poll::Ready(Some(val)) => {
+                self.gauge.dec();
+                Poll::Ready(Some(val))
+            }
+            s => s,
+        }
+    }
+}
+
+impl<T> Unpin for Receiver<T> {}
+
+impl<T> Sender<T> {
+    /// Sends a value, waiting until there is capacity.
+    /// Increments the gauge in case of a successful `send`.
+    pub async fn send(&self, value: T) -> Result<(), SendError<T>> {
+        (*self)
+            .inner
+            .send(value)
+            .inspect(|_| self.gauge.inc())
+            .await
+    }
+
+    /// Completes when the receiver has dropped.
+    pub async fn closed(&self) {
+        self.inner.closed().await
+    }
+
+    /// Attempts to immediately send a message on this `Sender`
+    /// Increments the gauge in case of a successful `try_send`.
+    pub fn try_send(&self, message: T) -> Result<(), TrySendError<T>> {
+        (*self)
+            .inner
+            .try_send(message)
+            // remove this unsightly hack once https://github.com/rust-lang/rust/issues/91345 is resolved
```

Comment:
Probably this is still not relevant as issue seems to be closed.

Reply:
I'm confused: this link works and is still current in the Rust documentation.